skyverge/wc-plugin-framework
# Adds a callback to AJAX requests to log script errors

## Summary

Adds an action callback that listens to AJAX requests to log a script error.

#### Story: [ch33855](https://app.clubhouse.io/skyverge/story/33855/add-an-ajax-callback-for-adding-a-debug-log-record-when-loading-of-a-payment-gateway-javascript-handler-fails)
#### PR: #470 

## Additional details

Unfortunately I've hit a small issue with this that prompted me to change the tasks implementation as detailed in the story. 

Seems that for at least the Payment Form handler, the callbacks we originally wanted to include in the `Frontend\Script_Handler` would not trigger early enough to catch the AJAX request and log the error. The Payment Form handler is output in frontend during template output time, so a bit too late and I could not see a way to include the AJAX handling directly inside `Frontend\Script_Handler`  and initializing the Payment Form early enough without running into risks of having certain hooks firing twice inside that handler, or other unnecessary refactorings. Instead, I placed the AJAX request callback inside the gateway main file, along with other log handling functions. This seems to work. 

Other details in this [Slack thread](https://skyverge.slack.com/archives/CV25U34GZ/p1585597377017300).

This issue was a bit difficult to catch early on as timing issues may not be easy to be seen, especially within abstractions presented by the framework.

## QA

You can test this by attaching the PR to a gateway plugin, then reach the screens where the script would appear. If there's an error (you can purposely trigger one), then an AJAX request would be triggered. If the callback has registered early enough, the issue would be logged in the plugin's main log.

Example log result you should see:

![image](https://user-images.githubusercontent.com/1227930/77982448-f4c5ed00-733e-11ea-8422-63b4dd0c9f2a.png)
